storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,766 @@
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleKillAndRelaunch(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
+        if (dynamicState.container.areAllProcessesDead()) {
+            if (equivalent(dynamicState.newAssignment, dynamicState.currentAssignment)) {
+                dynamicState.container.cleanUpForRestart();
+                dynamicState.container.relaunch();
+                return dynamicState.withState(MachineState.WAITING_FOR_WORKER_START);
+            }
+            //Scheduling changed after we killed all of the processes
+            return prepareForNewAssignmentOnEmptySlot(cleanupCurrentContainer(dynamicState, staticState, null), staticState);
+        }
+        //The child processes typically exit in < 1 sec.  If 2 mins later they are still around something is wrong
+        if ((Time.currentTimeMillis() - dynamicState.startTime) > 120000) {
+            throw new RuntimeException("Not all processes in " + dynamicState.container + " exited after 120 seconds");
+        }
+        dynamicState.container.forceKill();
+        Time.sleep(staticState.killSleepMs);
+        return dynamicState;
+    }
+
+    /**
+     * State Transitions for WAITING_FOR_WORKER_START state.
+     * PRECONDITION: container != null && currentAssignment != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleWaitingForWorkerStart(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
+        LSWorkerHeartbeat hb = dynamicState.container.readHeartbeat();
+        if (hb != null) {
+            long hbAgeMs = (Time.currentTimeSecs() - hb.get_time_secs()) * 1000;
+            if (hbAgeMs <= staticState.hbTimeoutMs) {
+                return dynamicState.withState(MachineState.RUNNING);
+            }
+        }
+        
+        if (!equivalent(dynamicState.newAssignment, dynamicState.currentAssignment)) {
+            //We were rescheduled while waiting for the worker to come up
+            return Slot.killContainerForChangedAssignment(dynamicState, staticState);
+        }
+        
+        long timeDiffms = (Time.currentTimeMillis() - dynamicState.startTime);
+        if (timeDiffms > staticState.firstHbTimeoutMs) {
+            LOG.warn("SLOT {}: Container {} failed to launch in {} ms.", staticState.port, dynamicState.container, staticState.firstHbTimeoutMs);
+            dynamicState.container.kill();
+            Time.sleep(staticState.killSleepMs);
+            return dynamicState.withState(MachineState.KILL_AND_RELAUNCH);
```

Comment:
We should just call `killAndRelaunchContainer` here. In this transition, there should not be any profiler actions as it has not come up yet, so that extra setup is a NOOP.